apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/strategy/ApplyNovaTemplateOptionsCreateNodesWithGroupEncodedIntoNameThenAddToSet.java

```diff
@@ -95,63 +91,73 @@ protected ApplyNovaTemplateOptionsCreateNodesWithGroupEncodedIntoNameThenAddToSe
    public Map<?, ListenableFuture<Void>> execute(String group, int count, Template template, Set<NodeMetadata> goodNodes,
             Map<NodeMetadata, Exception> badNodes, Multimap<NodeMetadata, CustomizationResponse> customizationResponses) {
 
-      Template mutableTemplate = template.clone();
-
-      NovaTemplateOptions templateOptions = NovaTemplateOptions.class.cast(mutableTemplate.getOptions());
-
-      assert template.getOptions().equals(templateOptions) : "options didn't clone properly";
-
-      String region = mutableTemplate.getLocation().getId();
-      ImmutableList.Builder<String> tagsBuilder = ImmutableList.builder();
+      NovaTemplateOptions templateOptions = NovaTemplateOptions.class.cast(template.getOptions());
+      String region = template.getLocation().getId();
 
       if (templateOptions.shouldAutoAssignFloatingIp()) {
          checkArgument(novaApi.getFloatingIPApi(region).isPresent(),
                  "Floating IPs are required by options, but the extension is not available! options: %s",
                  templateOptions);
       }
-
-      boolean keyPairExtensionPresent = novaApi.getKeyPairApi(region).isPresent();
       if (templateOptions.shouldGenerateKeyPair()) {
-         checkArgument(keyPairExtensionPresent,
-                  "Key Pairs are required by options, but the extension is not available! options: %s", templateOptions);
-         KeyPair keyPair = keyPairCache.getUnchecked(RegionAndName.fromRegionAndName(region, namingConvention.create()
-                  .sharedNameForGroup(group)));
-         keyPairCache.asMap().put(RegionAndName.fromRegionAndName(region, keyPair.getName()), keyPair);
-         templateOptions.keyPairName(keyPair.getName());
-         tagsBuilder.add(JCLOUDS_KP);
-      } else if (templateOptions.getKeyPairName() != null) {
-         checkArgument(keyPairExtensionPresent,
-                  "Key Pairs are required by options, but the extension is not available! options: %s", templateOptions);
-         if (templateOptions.getLoginPrivateKey() != null) {
-            String pem = templateOptions.getLoginPrivateKey();
-            KeyPair keyPair = KeyPair.builder().name(templateOptions.getKeyPairName())
-                     .fingerprint(fingerprintPrivateKey(pem)).privateKey(pem).build();
-            keyPairCache.asMap().put(RegionAndName.fromRegionAndName(region, keyPair.getName()), keyPair);
+         checkArgument(novaApi.getKeyPairApi(region).isPresent(),
+                 "Key Pairs are required by options, but the extension is not available! options: %s", templateOptions);
+      }
+      if (templateOptions.shouldGenerateKeyPair()) {
```

Comment:
Should we check instead if the user configured inbound ports?
I see that the groups are just passed through to the server options, so n need for the extension api if the user only sets the security group names?